File: src/offer35.Copy_Complex_List/copy_complex_list.cpp
/*
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(NULL), random(NULL) {
    }
};
*/
class Solution1 {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if (pHead == nullptr)
            return nullptr;
        RandomListNode* head = pHead;
        while (head != nullptr){
            RandomListNode* node = new RandomListNode(head->label);
            node->next = head->next;
            head->next = node;
            head = node->next;
        }
        head = pHead;
        while (head != nullptr){
            if (head->random){
                head->next->random = head->random->next;
            }
            head = head->next->next;
        }
        head = pHead;
        RandomListNode* newhead = pHead->next;
        RandomListNode* newNode = pHead->next;
        while (head != nullptr){
            head->next = head->next->next;
            if (newhead->next)
                newhead->next = newhead->next->next;
            head = head->next;
            newhead = newhead->next;
        }
        return newNode;
    }

};

/*
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(NULL), random(NULL) {
    }
};
*/
#include <unordered_map>
class Solution2 {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if (pHead == nullptr)
            return nullptr;
        unordered_map<RandomListNode*, RandomListNode*> hash;
        RandomListNode* tmp = pHead;
        while(tmp != nullptr){
            hash[tmp] = new RandomListNode(tmp->label);
            tmp = tmp->next;
        }
        tmp = pHead;
        while(tmp != nullptr){
            if (tmp->next)
                hash[tmp]->next = hash[tmp->next];
            if (tmp->random)
                hash[tmp]->random = hash[tmp->random];
            tmp = tmp->next;
        }
        return hash[pHead];
    }

};

/*
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(NULL), random(NULL) {
    }
};
*/
#include <unordered_map>
class Solution3 {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if (pHead == nullptr)
            return nullptr;
        unordered_map<RandomListNode*, RandomListNode*> hash;
        return dfs(pHead, hash);
    }
    RandomListNode* dfs(RandomListNode* node, unordered_map<RandomListNode*, RandomListNode*>& hash)
    {
        if (node == nullptr)
            return nullptr;
        if (hash.count(node))
            return hash[node];
        RandomListNode* newnode = new RandomListNode(node->label);
        hash[node] = newnode;
        hash[node]->next = dfs(node->next, hash);
        hash[node]->random = dfs(node->random, hash);
        return newnode;
    }
};

/*
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(NULL), random(NULL) {
    }
};
*/
#include <unordered_map>
class Solution4 {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if (pHead == nullptr)
            return nullptr;
        return bfs(pHead);
    }
    RandomListNode* bfs(RandomListNode* node)
    {
        if (node == nullptr)
            return nullptr;
        unordered_map<RandomListNode*, RandomListNode*> hash;
        hash[node] = new RandomListNode(node->label);
        queue<RandomListNode*> que;
        que.push(node);
        while(!que.empty()){
            int len = que.size();
            for(int i = 0; i < len; i++){
                RandomListNode* tmp = que.front();
                que.pop();
                if (tmp->next){
                    if (hash.count(tmp->next) == 0){
                        hash[tmp->next] = new RandomListNode(tmp->next->label);
                        que.push(tmp->next);
                    }
                    hash[tmp]->next = hash[tmp->next];
                }
                if (tmp->random){
                    if (hash.count(tmp->random) == 0){
                        hash[tmp->random] = new RandomListNode(tmp->random->label);
                        que.push(tmp->random);
                    }
                    hash[tmp]->random = hash[tmp->random];
                }
            }
        }
        return hash[node];
    }
};

File: src/offer35.Copy_Complex_List/copy_complex_list.java
/*
public class RandomListNode {
    int label;
    RandomListNode next = null;
    RandomListNode random = null;

    RandomListNode(int label) {
        this.label = label;
    }
}
*/
public class Solution {
    public RandomListNode Clone(RandomListNode pHead)
    {
        if(pHead == null){
            return null;
        }
        RandomListNode cur = pHead;
        while(cur != null){
            RandomListNode node = new RandomListNode(cur.label);
            node.next = cur.next;
            cur.next = node;
            cur = node.next;
        }
        cur = pHead;
        while(cur != null){
            if(cur.random !=null){                        //!!!!!!!!!!!!!!
                cur.next.random = cur.random.next;
            }
            cur = cur.next.next;
        }
        cur = pHead;
        RandomListNode newHead = cur.next;
        RandomListNode tmp = newHead;
        while(tmp != null){
            cur.next = cur.next.next;
            if(tmp.next != null){                          //!!!!!!!!!!!!!!!!
                tmp.next = tmp.next.next;
            }
            tmp = tmp.next;
            cur = cur.next;
        }
        return newHead;
    }
}

/*
public class RandomListNode {
    int label;
    RandomListNode next = null;
    RandomListNode random = null;

    RandomListNode(int label) {
        this.label = label;
    }
}
*/
import java.util.*;
public class Solution {
    public RandomListNode Clone(RandomListNode pHead)
    {
        if(pHead == null){
            return null;
        }
        HashMap<RandomListNode,RandomListNode> map = new HashMap<>();
        RandomListNode tmp = pHead;
        while(tmp != null){
            map.put(tmp, new RandomListNode(tmp.label));
            tmp = tmp.next;
        }
        tmp = pHead;
        while(tmp != null){
            if(tmp.next != null){
                map.get(tmp).next = map.get(tmp.next);
            }
            if(tmp.random != null){
                map.get(tmp).random = map.get(tmp.random);
            }
            tmp = tmp.next;
        }
        return map.get(pHead);
    }
}

/*
public class RandomListNode {
    int label;
    RandomListNode next = null;
    RandomListNode random = null;

    RandomListNode(int label) {
        this.label = label;
    }
}
*/
import java.util.*;
public class Solution {
    public RandomListNode Clone(RandomListNode pHead)
    {
        if(pHead == null){
            return null;
        }
        HashMap<RandomListNode, RandomListNode> map = new HashMap<>();
        return dfs(pHead, map);
    }
    private RandomListNode dfs(RandomListNode node, HashMap<RandomListNode, RandomListNode> map){
        if(node == null){
            return null;
        }
        if(map.containsKey(node)){
            return map.get(node);
        }
        RandomListNode newnode = new RandomListNode(node.label);
        map.put(node, newnode);
        map.get(node).next = dfs(node.next, map);                    //!!!!!!!!!!!!
        map.get(node).random = dfs(node.random, map);                 //!!!!!!!!!!!!
        return newnode;
    }
}

//
/*
public class RandomListNode {
    int label;
    RandomListNode next = null;
    RandomListNode random = null;

    RandomListNode(int label) {
        this.label = label;
    }
}
*/
import java.util.*;
public class Solution {
    public RandomListNode Clone(RandomListNode pHead)
    {
        if(pHead == null){
            return null;
        }
        Queue<RandomListNode> q = new LinkedList<>();
        HashMap<RandomListNode, RandomListNode> map = new HashMap<>();
        map.put(pHead, new RandomListNode(pHead.label));
        q.offer(pHead);
        while(!q.isEmpty()){
            int len = q.size();
            while(len > 0){
                RandomListNode cur = q.poll();
                if(cur.next != null){
                    if(!map.containsKey(cur.next)){
                        map.put(cur.next, new RandomListNode(cur.next.label));
                        q.offer(cur.next);
                    }
                    map.get(cur).next = map.get(cur.next);
                }
                if(cur.random != null){
                    if(!map.containsKey(cur.random)){
                        map.put(cur.random, new RandomListNode(cur.random.label));
                        q.offer(cur.random);
                    }
                    map.get(cur).random = map.get(cur.random);
                }
                len--;
            }
        }
        return map.get(pHead);
    }
}